### PR TITLE
[FIX] ir_model_fields: properly unlink manual fields

### PR DESCRIPTION
**Pull request by Elkasitu**
Before this commit:

* Install module A that implements model M and field X.

* Create a new custom field Y whose reference model is not M, implemented in
any module other than A.

* Make this field either a one2many, related or compute field depending
on field X.

* Uninstall module A.

=> The registry crashes fatally, the only way to recover from this is to
log directly into the postgres database and manually eliminate the
infringing ir.model.fields record.

The reasons are multiple:

* When creating custom fields that are relational, char fields are used
to hold data like model name, field name, instead of properly storing
the related field, which would let postgres handle the unlinking
properly.

* Despite this, the ir.model.fields method _prepare_update tries to make
up for this by verifying that the ir.model.fields being unlinked during
an uninstallation are not needed elsewhere and drops them, however it
does so naively: It assumes that the "problematic" field will be defined
within the module's scope, i.e. its model reference will be a model
defined by the module being uninstalled, this isn't always the case.

This is solved by getting a recordset of candidates, these are
ir.model.fields records that could be potentially troublesome for the
current uninstallation and should be deleted (ideally they should
restrict the uninstallation, but this is currently not possible).

These candidates are defined by being ir.model.fields records that are
custom (state = 'manual') and are of ttype 'one2many' or are computed or
related.

We compare each candidate and the current ir.model.fields being unlinked
and verify if they're related or not through a heuristic, if it's the
case then we unlink the candidate, if not we do nothing.

The unlinking of the candidate is only done in the case of an
uninstallation, if this is not the case an error will be raised instead
(existing behavior previous to this patch).

This commit is a backport + generalization of commit 4265026ee23ab71e8b199a9d8cbbd439409d7a03

Task-ID: 1856263

TODO:
- [x] Optimize heuristic
- [x] Write unit tests (set `MODULE_UNINSTALL_FLAG` in context)

### CODE REVIEW

File: odoo/addons/base/ir/ir_model.py [hunk 1]
```diff
@@ -546,23 +546,63 @@ def _prepare_update(self):
...
+                ('state', '=', 'manual'),
+            '|',
+                ('related', '!=', False),
+            '|',
+                ('depends', '!=', False),
+                ('ttype', '=', 'one2many'),
+        ])
+
+        def resolve_deps(candidate):
+            fields = []
```
Comment (Elkasitu) [inline odoo/addons/base/ir/ir_model.py]:
```python
model = self.env[candidate.model]
field = model._fields[candidate.name]
return [dep[1] for dep in field.resolve_deps(model)]
```

File: odoo/addons/base/ir/ir_model.py [hunk 2]
```diff
@@ -546,23 +546,63 @@ def _prepare_update(self):
...
+
+        for candidate in candidates:
+            for rec in self:
+                manual_unlink = False
+                model = self.env[rec.model]
+                field = model._fields[rec.name]
+                if candidate.ttype == 'one2many' and rec.ttype == 'many2one':
+                    if (candidate.relation == rec.model
+                            and candidate.relation_field == rec.name):
+                        if self._context.get(MODULE_UNINSTALL_FLAG):
```
Comment (Elkasitu) [inline odoo/addons/base/ir/ir_model.py]: put candidate to delete in a containertype

File: odoo/addons/base/ir/ir_model.py [hunk 3]
```diff
@@ -546,23 +547,54 @@ def _prepare_update(self):
             This method prevents the modification/deletion of many2one fields
             that have an inverse one2many, for instance.
         """
+        # Candidates are custom ir.model.fields that are either of type one2many or
+        # related or a compute field and will potentially be unlinked if they are a dependant
```
Comment (Elkasitu) [inline odoo/addons/base/ir/ir_model.py]: "that might depend on the fields corresponding to `self`"

File: odoo/addons/base/ir/ir_model.py [hunk 4]
```diff
@@ -546,23 +547,54 @@ def _prepare_update(self):
...
+                ('depends', '!=', False),
+                ('ttype', '=', 'one2many'),
+        ]) - self
+
+        def resolve_deps(candidate):
+            model = self.env[candidate.model]
+            field = model._fields[candidate.name]
+            return [dep[1] for dep in field.resolve_deps(model)]
+
+        relations_to_unlink = []
```
Comment (Elkasitu) [inline odoo/addons/base/ir/ir_model.py]: `failed_dependencies`

File: odoo/addons/base/ir/ir_model.py [hunk 5]
```diff
@@ -546,23 +547,54 @@ def _prepare_update(self):
...
+        for candidate in candidates:
+            for rec in self:
+                model = self.env[rec.model]
+                field = model._fields[rec.name]
+                if (candidate.ttype == 'one2many'
+                        and rec.ttype == 'many2one'
```
Comment (Elkasitu) [inline odoo/addons/base/ir/ir_model.py]: this condition is not necessary

File: odoo/addons/base/ir/ir_model.py [hunk 6]
```diff
@@ -546,23 +547,54 @@ def _prepare_update(self):
...
+                    dep_fields = resolve_deps(candidate)
+                    if field in dep_fields:
+                        relations_to_unlink.append((candidate, rec))
+
+        if not self._context.get(MODULE_UNINSTALL_FLAG) and relations_to_unlink:
+            # do not unlink candidate if not in uninstall mode
+            msg = _("The field '%s' cannot be removed because the field '%s' depends on it.")
+            raise UserError(msg % relations_to_unlink[0])
+        elif relations_to_unlink:
+            functools.reduce(lambda x, y: x | y, (rel[0] for rel in relations_to_unlink)).unlink()
```
Comment (Elkasitu) [inline odoo/addons/base/ir/ir_model.py]: `self.browse().union(*candidates_to_unlink])`

**Comment by Elkasitu** [comment]
With the new heuristic the increase in module uninstall time is negligible (~0.6s per 1000 custom fields)

File: odoo/addons/base/tests/test_ir_actions.py [hunk 7]
```diff
@@ -272,6 +273,49 @@ def test_create_custom(self):
...
+            'model_id': model.id,
+            'name': 'x_my_o2m',
+            'field_description': 'my_o2m',
+            'ttype': 'one2many',
+            'relation': self.COMODEL,
+            'relation_field': m2o_field.name,
+        })
+
+        m2o_field.unlink()
+        self.assertFalse(o2m_field.exists())
```
Comment (rco-odoo) [inline odoo/addons/base/tests/test_ir_actions.py]:
```python
# normal mode
with self.assertRaises(UserError):
    m2o_field.unlink()

# uninstall mode
m2o_field.with_context(_force_unlink=True).unlink()
self.assertFalse(o2m_field.exists())
```

**Pull request merged**